examples/ecs/indexing.rs

```diff
@@ -0,0 +1,205 @@
+//! This example demonstrates how to access `Component` data through an `Index`.
+#![allow(clippy::type_complexity)]
+
+use bevy::{
+    ecs::{
+        component::Tick,
+        system::{SystemChangeTick, SystemParam},
+    },
+    prelude::*,
+    utils::{HashMap, HashSet},
+};
+use bevy_internal::ecs::query::ReadOnlyWorldQuery;
+use std::{hash::Hash, marker::PhantomData};
+
+pub trait Indexer {
+    type Input: Component;
+    type Index: Hash + Eq + Clone + 'static;
+
+    fn index(input: &Self::Input) -> Self::Index;
+}
+
+pub struct SimpleIndexer<T>(PhantomData<T>);
+
+impl<T> Indexer for SimpleIndexer<T> where T: Component + Hash + Eq + Clone {
+    type Input = T;
+
+    type Index = T;
+
+    fn index(input: &Self::Input) -> Self::Index {
+        input.clone()
+    }
+}
+
+#[derive(Resource)]
+struct IndexBacking<T, F = (), I = SimpleIndexer<T>> {
+    forward: HashMap<T, HashSet<Entity>>,
+    reverse: HashMap<Entity, T>,
```

Comment:
It would be nice if this didn't require duplicating the `T`s. For example by using a single continuous buffer for the `T`s and then storing indices into that buffer in two `HashTable`s, kinda like how [`IndexMap`](https://docs.rs/indexmap/latest/indexmap/) works.

This may also benefit from the ad-hoc hasher used by [`EntityHashMap`/`EntityHashSet`](https://github.com/bevyengine/bevy/blob/e67cfdf82b5726db4d449e9af31b865a5324aa19/crates/bevy_utils/src/lib.rs#L299-L303)

Reply:
it would definitely make sense to use EntityHashMap/EntityHashSet here

Reply:
> It would be nice if this didn't require duplicating the `T`s. For example by using a single continuous buffer for the `T`s and then storing indices into that buffer in two `HashTable`s, kinda like how [`IndexMap`](https://docs.rs/indexmap/latest/indexmap/) works.

That would be nice, but I'm not certain how to get a truly bidirectional mapping without storing the index value in both halves. I would like to explore more general bijective storage options, but for now I know this will work, just not optimally.